### ISSUE TITLE
Proposal: napari.imshow, or changing view_* to return *both* the viewer and the layer

### ISSUE THREAD

**Issue by jni** [napari/napari#4693]
## 🚀 Feature / Motivation

Probably my most common napari usage involves quickly looking at an image, followed by some operations. If I want access to the viewer and the layer, I currently have two choices:

```python
viewer = napari.view_image(image)
image_layer = viewer.layers[0]
```

*or,*

```python
viewer = napari.Viewer()
image_layer = viewer.add_image(image)
```

The entire point of the `view_*` methods was to *avoid* a 2-line construct, and imho they have ultimately failed at that goal except in the simplest cases where we don't need access to the layer object. Typically what I find is that I think "ok *today* I won't need access and `view_` is perfect!", and then I invariably have to go hunt for a reference to that layer. 😂 

## Pitch

Instead I'd like to have a single construct:

```python
viewer, layer = napari.<function_name>(data)
```

I'm torn between modifying the `view_` methods to have this behaviour, or deprecating the view methods and just doing `napari.imshow()`.

## Alternatives

Python 3.8 sorta gives me my one-liner in:

```python
layer = (viewer := napari.view_image(image)).layers[0]
```

But that feels like an empty victory. 😂 

What do people think? Since this is a breaking API change, (a) I think we should aim for it to happen in 0.5, and (b) I'd like to gather a bit of feedback here before proposing a NAP.

**Comment by kevinyamauchi** [comment]
> layer = (viewer := napari.view_image(image)).layers[0]

🤣 

> What do people think?

My preference would be to modify the existing `napari.view_*` methods to return both the image and viewer because I like the names and it's a breaking change anyway, so I don't think it's too bad to change the return signature. I suppose the downside is that it's more difficult to deprecate slowly/give a warning, but if we do it with 0.5.0 and include it in the "migration guide", I think it's not so bad.

I can see `imshow` for images, but I think it's good for the names to have a pattern and I'm not sure that `shapeshow` and `tracksshow` have the same ring to them.

**Comment by Czaki** [comment]
add argument `return_layer: bool = False` ?

**Comment by jni** [comment]
> I can see imshow for images, but I think it's good for the names to have a pattern and I'm not sure that shapeshow and tracksshow have the same ring to them.

I agree regarding the pattern, but personally I think of images as the first-class layer in napari, with the other ones being adjacent to it, so I'd be ok with `imshow` being the *only* view function, with others requiring multiline.

Additionally, I think the key is for the other names to be abbreviated as well: shpshow, ptshow, segshow/labshow, vecshow, trkshow, srfshow. I quite like them tbh! 😃 

> add argument `return_layer: bool = False` ?

In skimage we've decided against this going forward. Although it's common in Python code, it makes typing a pain, as now you have value-dependent output type for that function. So I prefer always returning the same type of thing.

**Comment by Czaki** [comment]
> In skimage we've decided against this going forward. Although it's common in Python code, it makes typing a pain, as now you have value-dependent output type for that function. So I prefer always returning the same type of thing.

`Literal[False]` and `overload` are friends that allow handling this for mypy. But I see potential problems. 
But it still could be used in transition time (first warning and behave as the value set to `False` and then breaking change).

**Comment by tlambert03** [comment]
> I'm torn between modifying the view_ methods to have this behaviour, or deprecating the view methods and just doing napari.imshow().

I'd actually be ok with this, as I rarely use `view_<>`.  but it's a pretty big break, since these are used _everywhere_.

I'd also be ok with changing the return type to a tuple, but agree that that deprecation would be much harder/more invasive than a name change (hard to do in one step).

I also dislike the return type changing based on `return_layer` argument.  It's true that overloads are there for that, but overloads and literals are equally annoying and frail

**Comment by sofroniewn** [comment]
Picking up from https://github.com/napari/napari/pull/4858#issuecomment-1200505333

> add a napari.imshow() convenience function that returns a Viewer and a layer list (potentially with just a single element). See https://github.com/napari/napari/issues/4693. This function could optionally take a viewer= kwarg and use an existing viewer if needed, or create one if None.

```python
viewer, layers = napari.imshow(data, multichannel=False,  viewer=None, **kwargs)
```

where `layers` is a `list` (not a `LayerList`) of any added `Layer` objects - will be `1` if `multichannel=False` and will be more if `multichannel=True`. The other `**kwargs` are the same as for `add_image`.

Is that how you see it @jni?

If so we should probably have one PR to add `imshow` then one PR to deprecate `view_*` and then return to #4858 after both of those have merged. @tlambert03 is that how you see it too?

**Comment by tlambert03** [comment]
i think that works for me if everyone else likes it!  (but `multichannel` needs to be an int right?  to know which axis to split?)

**Comment by sofroniewn** [comment]
> multichannel needs to be an int right? to know which axis to split?

yup yup! was moving too quick - that's right

**Comment by sofroniewn** [comment]
If we have consensus @jni @tlambert03 is [this](https://github.com/napari/napari/issues/4693#issuecomment-1200509018) maybe a good thing for @kcpevey to pick up when possible @potating-potato and get this issue closed? Doing this will then unblock #4858 which we want to get in asap I think

**Comment by kcpevey** [comment]
Should I implement this similar to how the `view_` and `add_` methods are done? 

I've started down that path, but its a pretty tricky implementation - little bits here and there in many files, no explicitly named methods so its hard to search/track. I can keep going, but just wanted to confirm this direction before proceeding.

**Comment by tlambert03** [comment]
I'd say don't start with auto generation.  Just make a regular python method, and if we later need/want to do any auto generation we can do so in a refactor

**Comment by kcpevey** [comment]
JUST figured it out - I needed to add it as an attr in both `view_layers.py` _and_ `napari.__init__.py`. 

I have a single image working, but I need to handle multichannnel and an existing viewer.

**Comment by kcpevey** [comment]
If I didn't do it via auto-generation, where would the method be? Still in `view_layers.py`? But then I'd still need to add it as an attr like the `add_` ad `view_` methods because of [this comment](https://github.com/napari/napari/blob/main/napari/__init__.py#L42-L49)?

**Comment by tlambert03** [comment]
> where would the method be

since the proposal is for a top level `napari.imshow` method, you can put it pretty much anywhere you want, as long as you export it in `__init__.py` by adding it to the `_submod_attrs` dict in `napari/__init__.py`.  `view_layers.py` would be an ok place to put it, but you could also put it in something like `napari.utils._imshow` ... i don't have a strong opinion on where the method itself lives, only where we export it.

**Comment by kcpevey** [comment]
I've opened up a PR, but it still needs work. I thought it would be best to continue the discussion over there where we will have code to look at :)

**Comment by psobolewskiPhD** [comment]
https://github.com/napari/napari/pull/4928 has been merged, can we close this @jni?

**Comment by jni** [comment]
There's still some decisions to make that are effectively being tracked by this issue:

- Do we deprecate the `view_` functions?
- Do we add `ptshow` and similar?

I think yes to deprecating the `view_` functions (and we close this after the deprecations are added), no (for now) to the non-image `show` functions. (And we open a new issue if people love `imshow` so much that they miss the sister functions.)

**Comment by jni** [comment]
I'm going to close this and make a different issue for the `view_*` deprecation, since this one is hard to grok at a glance given the mostly-completed part.
